### ISSUE TITLE
Type guard for undefined not working for intersection type

### ISSUE THREAD

**Issue by cvuorinen** [Microsoft/TypeScript#28079]
I am using an empty enum as intersection with string to augment a nominal type (according to these instructions https://basarat.gitbooks.io/typescript/docs/tips/nominalTyping.html#using-enums). Otherwise it's working correctly, but I'm getting errors from nullable properties even though there is an if check to make sure it's not undefined or null.

**TypeScript Version:**  3.2.0-dev.20181023

**Search Terms:** "intersection nullable guard" "intersection undefined guard" "undefined nominal guard" "guard undefined"

**Code**

```ts
// Regular type alias (working correctly)
type UuidString = string

type DataWithString = {
    id?: UuidString
};

function doSomethingWithString(id: UuidString) {
    // ...
}

function handleDataWithString(data: DataWithString) {
    if (data.id) { // if condition correctly acts as guard against undefined
        doSomethingWithString(data.id) // No error here
    }
}

// nominal type for UUID
// according to https://basarat.gitbooks.io/typescript/docs/tips/nominalTyping.html#using-enums
enum UuidType { }
type Uuid = UuidType & string

type DataWithUuid = {
    id?: Uuid
}

function doSomethingWithUuid(id: Uuid) {
    // ...
}

function handleUuid(data: DataWithUuid) {
    if (data.id) { // if condition not working as guard against undefined
        doSomethingWithUuid(data.id) // <-- Error
    }
}
```

**Expected behavior:**
*no errors*

**Actual behavior:**
```
test.ts:33:29 - error TS2345: Argument of type 'undefined' is not assignable to parameter of type 'Uuid'.
  Type 'undefined' is not assignable to type 'UuidType'.

33         doSomethingWithUuid(data.id) // <-- Error
                               ~~~~~~~
```

[**Playground Link:**](https://www.typescriptlang.org/play/index.html#src=type%20UuidString%20%3D%20string%0D%0A%0D%0Atype%20DataWithString%20%3D%20%7B%0D%0A%20%20%20%20id%3F%3A%20UuidString%0D%0A%7D%3B%0D%0A%0D%0Afunction%20doSomethingWithString(id%3A%20UuidString)%20%7B%0D%0A%20%20%20%20%2F%2F%20...%0D%0A%7D%0D%0A%0D%0Afunction%20handleDataWithString(data%3A%20DataWithString)%20%7B%0D%0A%20%20%20%20if%20(data.id)%20%7B%0D%0A%20%20%20%20%20%20%20%20doSomethingWithString(data.id)%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0A%2F%2F%20nominal%20type%20for%20UUID%0D%0A%2F%2F%20according%20to%20https%3A%2F%2Fbasarat.gitbooks.io%2Ftypescript%2Fdocs%2Ftips%2FnominalTyping.html%23using-enums%0D%0Aenum%20UuidType%20%7B%20%7D%0D%0Atype%20Uuid%20%3D%20UuidType%20%26%20string%0D%0A%0D%0Atype%20DataWithUuid%20%3D%20%7B%0D%0A%20%20%20%20id%3F%3A%20Uuid%0D%0A%7D%0D%0A%0D%0Afunction%20doSomethingWithUuid(id%3A%20Uuid)%20%7B%0D%0A%20%20%20%20%2F%2F%20...%0D%0A%7D%0D%0A%0D%0Afunction%20handleUuid(data%3A%20DataWithUuid)%20%7B%0D%0A%20%20%20%20if%20(data.id)%20%7B%0D%0A%20%20%20%20%20%20%20%20doSomethingWithUuid(data.id)%0D%0A%20%20%20%20%7D%0D%0A%7D) (+ select strictNullChecks from options) 

**Related Issues:**

**Comment by andy-ms** [comment]
`enum` by default is a number type. So `UuidType & string` is something that is both a number and a string, which can never happen. When you write `id?: Uuid`, that generates a property whose type is `Uuid | undefined`. When TypeScript sees a union type it will try to simplify it, such as by removing impossible members; `Uuid` is impossible so the union simplifies to `undefined`. The `if` test won't work because `data.id` doesn't have any non-undefined types to narrow to.

**Comment by jack-williams** [comment]
I think another way to interpret the behaviour, according to this [comment](
https://github.com/Microsoft/TypeScript/issues/21998#issuecomment-373162494), is that the an enum type is the union type of its members. As the enum is empty its type is the empty union, which is the same as the `never` type. So `never & string` reduces to `never`. Either way the result is the same as described by Andy.

**Comment by svieira** [comment]
Can be made to work by switching the enum to a string enum to get the nominality:

```ts
enum UuidUniverse { UuidType = '' }  // <-- Now a string enum
type Uuid = UuidUniverse.UuidType & string

type DataWithUuid = {
    id?: Uuid
}

function doSomethingWithUuid(id: Uuid) {
    // ...
}

function handleUuid(data: DataWithUuid) {
    if (data.id) { // if condition now working as guard against undefined
        doSomethingWithUuid(data.id) // <-- Works
    }
}
```

Symbol-driven branding doesn't work (for the same reasons as the int-based enums don't work) as Andy has already pointed out.  Literal-driven branding does work too:

```ts
interface Nominal<T/*: must be an enum or a literal */> {
  'nominal structural brand': T
}
enum UuidType {}
type Uuid = string & Nominal<UuidType>;

// ... etc. ...
```

**Comment by andy-ms** [comment]
```ts
enum UuidUniverse { UuidType = '' }  // <-- Now a string enum
type Uuid = UuidUniverse.UuidType & string
```

If you hover over it you'll see `Uuid = UuidUniverse`, because:
* `UuidUniverse.UuidType` is already a string, so `& string` has no effect
* `UuidUniverse` has only one element and so is equivalent to that element.

**Comment by cvuorinen** [comment]
Thanks @svieira, just came here to post that exact same thing as I figured I'll try it based on andys comment on them being different type. But now reading again what @andy-ms says, it means 

```typescript
enum UuidUniverse { UuidType = '' }  // <-- Now a string enum
type Uuid = UuidUniverse.UuidType & string
```

is equivalent to just 

```typescript
enum Uuid { UuidType = '' }
```

Not sure if there are any downsides to using just that, it seems to work correctly for my use case. Which is that we get data from another server as JSON and deserialize it, we have typings for the data and there are some id's that are UUIDs and then some other IDs that are string but not UUID. And we had few bugs where we passed the wrong type of IDs to functions that expected an UUID (as we previously just had `type Uuid = string`). We do not create those IDs in our codebase, they just flow through. So with this kind of typing we can catch those bugs since a value typed as Uuid can be used anywhere a string is expected but plain string cannot be used where Uuid is expected.

Another solution is to use the original code I posted and just add `!` in the line that gives the error, e.g. `doSomethingWithUuid(data.id!)`. Not that big a deal I guess. Or anyone have any other suggestions to handling this kind of situation?
